src/shogun/io/ARFFFile.h

```diff
@@ -0,0 +1,713 @@
+
+			if (skip_first && !m_stream->eof())
+				func();
+
+			while (!m_state && !m_file_done)
+			{
+				consume_line(func);
+			}
+			if (!check_func())
+			{
+				SG_SERROR(
+				    "Parsing error on line %d: %s\n", m_line_number,
+				    m_current_line.c_str());
+			}
+		}
+
+		/**
+		 * Function called by process_chunk to process a "chunk" line by line.
+		 * This function also checks if EOF has been reached.
+		 *
+		 * @tparam T type of processing function
+		 * @param func line processing function
+		 */
+		template <typename T>
+		void consume_line(T&& func)
+		{
+			if (m_stream->eof())
+			{
+				m_file_done = true;
+				return;
+			}
+			std::getline(*m_stream, m_current_line);
+			m_line_number++;
+			if (!arff_detail::is_blank(m_current_line))
+				func();
+		}
+#ifndef SWIG
+		/**
+		 * Checks if a token represented by a string
+		 * denotes a primitive type in the ARFF format
+		 * @param token the token to be checked
+		 * @return whether the token denotes a primitive type
+		 */
+		SG_FORCED_INLINE bool is_primitive_type(const std::string& token) const
+		    noexcept
+		{
+			return token.find_first_of("numeric") != std::string::npos ||
+			       token.find_first_of("integer") != std::string::npos ||
+			       token.find_first_of("real") != std::string::npos ||
+			       token.find_first_of("string") != std::string::npos;
+		}
+#endif // SWIG
+		template <typename ScalarType, typename CharType>
+		void reserve_vector_memory(
+		    size_t line_count,
+		    std::vector<variant<
+		        std::vector<ScalarType>,
+		        std::vector<std::basic_string<CharType>>>>& v);
+
+		/** the name of the attributes */
+		std::vector<std::string> m_attribute_names;
+
+		/** the input stream */
+		std::shared_ptr<std::istream> m_stream;
+		/** the scalar type used for parsing */
+		EPrimitiveType m_primitive_type;
+		/** the string underlying type used for parsing */
+		EPrimitiveType m_string_primitive_type;
+
+		/** internal line number counter for exceptions */
+		size_t m_line_number;
+		/** internal flag set true when string stream is EOF */
+		bool m_file_done;
+		/** internal state when set to true switches parsing rules */
+		bool m_state;
+		/** current row count of data */
+		size_t m_row_count;
+		/** the string after m_relation_string*/
+		std::string m_relation;
+		/** the string where comments are stored */
+		std::vector<std::string> m_comments;
+		/** the string representing the current line being parsed */
+		std::string m_current_line;
+		/** the attribute types in the order they are parsed */
+		std::vector<Attribute> m_attributes;
+		/** stores the date formats */
+		std::vector<std::string> m_date_formats;
+		/** the mapping of nominal attributes to their value */
+		std::vector<std::pair<std::string, std::vector<std::string>>>
+		    m_nominal_attributes;
+
+		/** the parsed features */
+		std::vector<std::shared_ptr<CFeatures>> m_features;
+	};
+
+	class ARFFSerializer
+	{
+	public:
+		ARFFSerializer(
```

Comment:
@karlnapf I had some time so I wrote this (pretty bad) serialiser.. Would it make sense to add a visitor pattern to SGObjects? Or at least for features?